### PR TITLE
CI updates and compile-warning fixes

### PR DESCRIPTION
**Pull request by dtebbs**
- use all cores in ci build
- build more configs (and enable all debug code)
- compile warnings as errors

### CODE REVIEW

File: libsnark/common/data_structures/integer_permutation.hpp [hunk 1]
```diff
@@ -31,6 +31,7 @@ class integer_permutation
     integer_permutation(const size_t size = 0);
     integer_permutation(const size_t min_element, const size_t max_element);
 
+    integer_permutation(const integer_permutation &other) = default;
```
Comment (AntoineRondelet) [inline libsnark/common/data_structures/integer_permutation.hpp:34]: Why is this one needed?

Comment (dtebbs) [reply libsnark/common/data_structures/integer_permutation.hpp:34]: Without this, the compiler gives a warning that an assignment operator has been defined, but a copy-constructor has not.

Comment (AntoineRondelet) [reply libsnark/common/data_structures/integer_permutation.hpp:34]: Ah I see, thanks

**Review by AntoineRondelet — APPROVED** [review]
LGTM

**Pull request merged**